text/0000-reloadable-process-types.md

```diff
@@ -0,0 +1,36 @@
+# Reloadable Process Types
+
+## Summary
+
+Provide a utility buildpack to support language buildpacks in creating reloadable process types for their run containers.
+
+## Motivation
+
+Usage of dev orchestration tools like Tilt and Skaffold are on the rise. These tools aim to make  developing distributed software that runs on Kubernetes easier.  Offering the capability to develop inside the container via live reload features.  With this, changes are streamed from the developer’s workstation into the running container whereupon the init process is restarted.  The dev orchestrator, of course, handles most of this by building the changes and copying them in the container but ultimately the running container needs the ability to restart its entrypoint process for the changes to take effect.
+
+Today buildpacks are unable to participate in these live reload development workflows but, moving forward, we would very much like them to be able to.
+
+## Detailed Explanation
+
+We propose creating a utility buildpack that packages the watchexec command line utility for restarting arbitrary processes.
+
+When enabled by a BP_LIVE_RELOAD_ENABLED=true environment variable, the language buildpacks may request the inclusion of watchexec, via the build plan.
```

Comment:
Here's what I'm taking away from this, please let me know if I'm missing anything:

This buildpack will participate (i.e. detection passes) if all of the following conditions are met:

- BP_LIVE_RELOAD_ENABLED=true is set
- A buildpack requests `watchexec` (or some other key) in the build plan.

The buildpack will do the following:

- Contributes `watchexec` to a layer with command on $PATH

What I'm unclear about is:

1. Will this buildpack be contributing any process types? or is that going to be the responsibility of another buildpack to do? For example, in Java the executable-jar buildpack contributes process types for running the executable JAR. Would the executable-jar buildpack have to do something different if `watchexec` is present? 

2. If this buildpack is responsible for providing `watchexec` in the build plan, what buildpack is responsible for requiring `watchexec` in the buildplan? Will this buildpack both provide & require it, which is fine? Or is some other buildpack supposed to require it?

If other buildpacks need to be changed to opt-in and participate with this buildpack, I think you need to add some more detail here. Specifically, what buildpacks will need to be modified, what changes will need to be made to those and what do the order groups need to be in the composite buildpacks so this all works correctly.

Reply:
> This buildpack will participate (i.e. detection passes) if all of the following conditions are met:
>
> BP_LIVE_RELOAD_ENABLED=true is set
> A buildpack requests watchexec (or some other key) in the build plan.

@dmikusa The way I am thinking about this the watchexec buildpack does not require `BP_LIVE_RELOAD_ENABLED=true` to be true to pass detect.  The buildpack will always provide watchexec and thus participate if a downstream buildpack requests watchexec via the build plan. The downstream buildpack on the other hand, makes a decision about whether to request and use watchexec based on the value of `BP_LIVE_RELOAD_ENABLED`.